apollo-router/src/spec/query/change.rs

```diff
@@ -132,158 +214,268 @@ impl<'a> QueryHashVisitor<'a> {
             }
             ExtendedType::Enum(e) => {
+                "^ENUM".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &e.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
 
+                "^ENUM_VALUE_LIST".hash(self);
                 for (value, def) in &e.values {
+                    "^VALUE".hash(self);
+
                     value.hash(self);
                     for directive in &def.directives {
                         self.hash_directive(directive);
                     }
                 }
+                "^ENUM_VALUE_LIST_END".hash(self);
             }
             ExtendedType::InputObject(o) => {
+                "^INPUT_OBJECT".hash(self);
+
+                "^DIRECTIVE_LIST".hash(self);
                 for directive in &o.directives {
                     self.hash_directive(&directive.node);
                 }
+                "^DIRECTIVE_LIST_END".hash(self);
 
+                "^FIELD_LIST".hash(self);
                 for (name, ty) in &o.fields {
-                    if ty.default_value.is_some() {
-                        name.hash(self);
-                        self.hash_input_value_definition(&ty.node)?;
-                    }
+                    "^NAME".hash(self);
+
+                    name.hash(self);
+
+                    "^ARGUMENT".hash(self);
+                    self.hash_input_value_definition(&ty.node)?;
                 }
+                "^FIELD_LIST_END".hash(self);
             }
         }
+        "^EXTENDED_TYPE-END".hash(self);
+
         Ok(())
     }
 
     fn hash_type(&mut self, t: &ast::Type) -> Result<(), BoxError> {
+        "^TYPE".hash(self);
+
         match t {
-            schema::Type::Named(name) => self.hash_type_by_name(name.as_str()),
+            schema::Type::Named(name) => self.hash_type_by_name(name.as_str())?,
             schema::Type::NonNullNamed(name) => {
                 "!".hash(self);
-                self.hash_type_by_name(name.as_str())
+                self.hash_type_by_name(name.as_str())?;
             }
             schema::Type::List(t) => {
                 "[]".hash(self);
-                self.hash_type(t)
+                self.hash_type(t)?;
             }
             schema::Type::NonNullList(t) => {
                 "[]!".hash(self);
-                self.hash_type(t)
+                self.hash_type(t)?;
             }
         }
+        "^TYPE-END".hash(self);
+        Ok(())
     }
 
     fn hash_field(
         &mut self,
         parent_type: String,
-        type_name: String,
         field_def: &FieldDefinition,
-        arguments: &[Node<Argument>],
+        node: &executable::Field,
     ) -> Result<(), BoxError> {
-        if self.hashed_fields.insert((parent_type.clone(), type_name)) {
-            self.hash_type_by_name(&parent_type)?;
+        "^FIELD".hash(self);
```

Comment:
I think it makes sense to hash field (parent type + field name) only once, similar to
https://github.com/apollographql/router/blob/0f8d90dace721e749fa3224579eee9ceda0e5aa4/apollo-router/src/spec/query/change.rs#L243-L248

Without it maybe be possible to send some expensive fields (e.g., a lot of arguments) bunch of times (intentionally or not).

Reply:
this was there initially but was part of the issue, because if a field was already seen once, it would not hash again both the field definition and the field usage in query, so if a field was present twice with different arguments, it would be incorrect. I think we can reintroduce it safely now

Reply:
done in 1598fbfa5, WDYT?

Reply:
@Geal 👍